core/src/main/scala/kafka/server/KafkaApis.scala

```diff
@@ -2384,68 +2385,116 @@ class KafkaApis(val requestChannel: RequestChannel,
+        for (topicPartition <- partitionsToAdd) {
+          if (!authorizedTopics.contains(topicPartition.topic))
+            unauthorizedTopicErrors += topicPartition -> Errors.TOPIC_AUTHORIZATION_FAILED
+          else if (!metadataCache.contains(topicPartition))
+            nonExistingTopicErrors += topicPartition -> Errors.UNKNOWN_TOPIC_OR_PARTITION
+          else
+            authorizedPartitions.add(topicPartition)
+        }
+
+        if (unauthorizedTopicErrors.nonEmpty || nonExistingTopicErrors.nonEmpty) {
+          // Any failed partition check causes the entire transaction to fail. We send the appropriate error codes for the
+          // partitions which failed, and an 'OPERATION_NOT_ATTEMPTED' error code for the partitions which succeeded
+          // the authorization check to indicate that they were not added to the transaction.
+          val partitionErrors = unauthorizedTopicErrors ++ nonExistingTopicErrors ++
+            authorizedPartitions.map(_ -> Errors.OPERATION_NOT_ATTEMPTED)
+          responses.add(AddPartitionsToTxnResponse.resultForTransaction(transactionalId, partitionErrors.asJava))
+          maybeSendResponse()
+        } else {
+          def sendResponseCallback(error: Errors): Unit = {
+            val finalError = {
+              if (version < 2 && error == Errors.PRODUCER_FENCED) {
                 // For older clients, they could not understand the new PRODUCER_FENCED error code,
                 // so we need to return the old INVALID_PRODUCER_EPOCH to have the same client handling logic.
                 Errors.INVALID_PRODUCER_EPOCH
               } else {
                 error
               }
-
-            val responseBody: AddPartitionsToTxnResponse = new AddPartitionsToTxnResponse(requestThrottleMs,
-              partitionsToAdd.map{tp => (tp, finalError)}.toMap.asJava)
-            trace(s"Completed $transactionalId's AddPartitionsToTxnRequest with partitions $partitionsToAdd: errors: $error from client ${request.header.clientId}")
-            responseBody
+            }
+            responses.synchronized {
+              responses.add(addPartitionsToTxnRequest.errorResponseForTransaction(transactionalId, finalError))
+            }
+            maybeSendResponse()
+          }
+          
+          def sendVerifyResponseCallback(errors: Map[TopicPartition, Errors]): Unit = {
+            responses.synchronized {
+              responses.add(AddPartitionsToTxnResponse.resultForTransaction(transactionalId, errors.asJava))
+            }
+            maybeSendResponse()
           }
 
-          requestHelper.sendResponseMaybeThrottle(request, createResponse)
+          txnCoordinator.handleAddPartitionsToTransaction(transactionalId,
+            transaction.producerId,
+            transaction.producerEpoch,
+            authorizedPartitions,
+            transaction.verifyOnly,
+            sendResponseCallback,
+            sendVerifyResponseCallback,
```

Comment:
I am not sure to understand why we need two callbacks. I find it weird in the first place. My understanding is that in both cases, we end up with an error per partition in the response so it seems to me that we could unify them, no?

Reply:
It is not the same. In one case we have a single error for all topic partitions, but in the other, we may have different errors per partition. (If we verify and have some that exist and others that don't)

Reply:
> It is not the same. In one case we have a single error for all topic partitions, but in the other, we may have different errors per partition. (If we verify and have some that exist and others that don't)

Yeah, I understand this. My point was that on the non-verify case, we also end up having an error per partition in the response (the same one for all of them).

> I had to keep the single error because this callback is used in the append path and I didn't want to change it.

Hum... I think that we could handle this in `handleAddPartitionsToTransaction` without changing the append path. Basically, when an error is return by the append, we could build a map with the error for all partitions.

I think that having two callbacks is a bad design choice here but that's just my humble opinion.

Reply:
Is the request to make a mapping of every topic partition to the same error? That doesn't seem super efficient, but I guess it makes things simpler in the parameters. We would also need to define a new callback in `handleAddPartitionsToTransaction` that calls the original callback (so we have two still)

I suppose this is doable though.

Reply:
I've changed to one callback, but I decided to go with a different approach where we either provide one error or the map. I think it would be annoying to convert the errors for all partitions and then potentially have to redo it to convert the error for v2 and below. Let me know if there is an issue with this.

Reply:
I am not sure whether I like it or not. Those two values in the callback may be error prone in the future. Reading `handleAddPartitionsToTransaction` again, the add and verify steps only have the code validation in common. How feasible would it be to actually add another method for the verification part instead of trying to do everything in `handleAddPartitionsToTransaction`? I suppose that we could extract the code part into a common method used by both.

Reply:
I suppose I can do that but it looked like a lot of code duplication. There wasn't a great way to extract out the commonalities but I can try again.